# Exploring eBay Car Sales Data

## Objective
In the project, we are going to clean the data and analyse the used car listings. 

The dataset can be found [here](https://data.world/data-society/used-cars-data). The dataset was originally scraped and uploaded to Kaggle by user [orgesleka](https://www.kaggle.com/orgesleka), but it is not avaliable on Kaggle anymore.

## The description of each column is as below:

- `dataCrawled`: When this ad was first crawled. All field-values are taken from this date.
- `name`: Name of the car.
- `seller`: Whether the seller is an individual or a seller.
- `offerType`: The type of listing. 
- `price`: The price on the ad to sell the car.
- `abtest`: Whether the listing is included in an A/B test.
- `vehicleType`: The vehicle Type.
- `yearOfRegistration`: The yar in which the car was first registered.
- `gearbox`: The transmission type.
- `powerPS`: The power of the car in PS.
- `model`: The car model name.
- `kilometer`: How many kilometers the car has driven.
- `monthOfRegistration`: The month in which the car was first registerd.
- `fuelType`: What type of fuel the car uses.
- `brand`: The brand of the car.
- `notRepairedDamage`: If the car has a damage which is not yet repaired.
- `dataCreated`: The date on which the eBay listing was created.
- `nrOfPictures`: The number of pictures in the ad.
- `postalCode`: The postal code for the location of the vehicle.
- `lastSeenOnline`: When the crawler saw this ad last online.


## Import and have a glimpse of the data

In [1]:
#importing libraries
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:
print(autos.info())
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

From the code above, we found some problmes:
1. The columns are mixed with Camelcase and Snakecase. (We use snakecase more often when analysing data)
2. There are some **missing values** in `vehicleType`, `gearbox`, `model`, `fuelType` and `notRepairedDamage`. 
3. Some values that could be numeric values are strings, such as `price`, and `odometer`.
4. As this dataset is from the German eBay website, some of the words are in German.

Therefore, we have to do further investigation in these values and fix or drop the data if needed before starting doing our analysis.

In [3]:
# Change column names from camelcase to snakcase.

print(autos.columns)
autos.columns = ['date_crawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'nrofpictures', 'postalcode',
       'last_seen']
autos.head()

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode            last_seen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

To make the column names more meaningfule, we changed the names from camelcase to snakecase. Also, we rename some column names. `yearOfRegistration` to `registration_year`, `monthOfRegistration` to `registration_month`, `notRepairedDamage` to `unrepaired_damage` and `dateCreated` to `ad_created`.

## First Part - Data Cleaning

In [4]:
autos.describe(include="all")

date_crawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

Investigating above table, I have some findings. 
1. `price` column can be converting from string to numeric values by removing "$" and ",". Likewise, by removing "km" and ",", `odometer` column can be converted from string to numeric values.
2. Columns `Seller` and `offertype` both have frequency of 49,999. This tells us that except for one value, almost all other values are the same. Therefore, we should remove these two columns.
3. In the column `nrofpictures`, there are only one value in this column, which is "0". We should also discard this column.

In [5]:
# Fixing price column
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)

# Fixing odometer column 
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

# Rename odometer to odometer_km
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)


We renamed `odometer` to `odometer_km` as km is an important information in the odometer columns.

In [6]:
# Removing unwanted columns
autos.drop(["seller","offertype","nrofpictures"], axis=1)

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price   abtest vehicletype  registration_year    gearbox  powerps  \
0       5000  control         bus               2004    manuell      158   
1       8500  control   limousine               1997  automatik      286   
2       8990     test   limousine               2009    manuell      102   
3       4350  control  kleinwagen               2007  automatik       71   
4       1350     test       kombi               2003    manuell        0   
...      ...      ...         ...                ...        ...      ...   
49995  24900  control   limousine               2011  automatik      239   
49996   1980  control      cabrio               1996    manuell       75   
49997  13200     test      cabrio               2014  automatik       69   
49998  22900  control       kombi               2013    manuell      150   
49999   1250  control   limousine               1996    manuell      101   

        model  odometer_km  registration_month fueltype       brand  \
0      andere       150000                   3      lpg     peugeot   
1         7er       150000                   6   benzin         bmw   
2        golf        70000                   7   benzin  volkswagen   
3      fortwo        70000                   6   benzin       smart   
4       focus       150000                   7   benzin        ford   
...       ...          ...                 ...      ...         ...   
49995      q5       100000                   1   diesel        audi   
49996   astra       150000                   5   benzin        opel   
49997     500         5000                  11   benzin        fiat   
49998      a3        40000                  11   diesel        audi   
49999  vectra       150000                   1   benzin        opel   

      unrepaired_damage           ad_created  postalcode            last_seen  
0                  nein  2016-03-26 00:00:00       79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00       71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00       35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00       33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00       39218  2016-04-01 14:38:50  
...                 ...                  ...         ...                  ...  
49995              nein  2016-03-27 00:00:00       82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00       44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00       73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00       35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00       45897  2016-04-06 21:18:48  

[50000 rows x 17 columns]

We drop "seller", "offertype" and "nrofpictures" columns because almost all values in theses columns are the same.

## Second Part - Exploring Data in Price and Odomeder columns.

In [7]:
#Explore price column
price_shape = autos["price"].shape
price_desc = autos["price"].describe()
price_value_counts = autos["price"].value_counts().sort_index(ascending=False)

print(price_shape)
print("\n")
print(price_desc)
print("\n")
print(price_value_counts)

(50000,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
            ... 
5              2
3              1
2              3
1            156
0           1421
Name: price, Length: 2357, dtype: int64


In [8]:
autos[autos["price"] > 400000].shape

(14, 20)

**Observations:**
- We defined price > 400000 as the outliers, thus we will exclude 14 values when analysing later.
- odometers does not contain outliers.


In [9]:
price_clean = autos[autos["price"].between(0, 400000)]
print(price_clean["price"].describe())

autos.loc[(autos["price"] < 0)]

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price, dtype: float64


Empty DataFrame
Columns: [date_crawled, name, seller, offertype, price, abtest, vehicletype, registration_year, gearbox, powerps, model, odometer_km, registration_month, fueltype, brand, unrepaired_damage, ad_created, nrofpictures, postalcode, last_seen]
Index: []

In [10]:
# Explore odometer_km column
odometer_km_shape = autos["odometer_km"].shape
odometer_km_desc = autos["odometer_km"].describe()
odometer_km_value_counts = autos["odometer_km"].value_counts(ascending=False)

print(odometer_km_shape)
print("\n")
print(odometer_km_desc)
print("\n")
print(odometer_km_value_counts)

(50000,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


**Observations:** 
- Odometer can be any value as long as its not negative. Therefore, there is no outliers in this column.

## Third Part - Explore Datetime
There are five columns that are datetime.
- `date_crawled`: added by the crawler (string column)
- `ad_created`: added by the crawler (string column)
- `last_seen`: from the website (string column)
- `registration_month`: from the website (integer column)
- `registration_year`: from the website (integer column)

In [11]:
# Have a quick look of the string columns
autos[["date_crawled", "ad_created", "last_seen"]][:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The data formate is timestamp (Y-m-d h:m:s). Let's check the range of the date in each of these string columns.

In [12]:
sort_index_date_crawled = autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False)
# normalize=True, so we can get the percentage 
# dropna=False, so we can see if there are missing data
# .sort_index(ascending=False), so we can check the date range in the column in a descending order.

top3_crawled_date = autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_values(ascending=False).head(3)

desc_crawled_date = autos["date_crawled"].str[:10].describe()

print(sort_index_date_crawled)
print("\n")
print(top3_crawled_date)
print("\n")
print(desc_crawled_date)

2016-04-07    0.00142
2016-04-06    0.00318
2016-04-05    0.01310
2016-04-04    0.03652
2016-04-03    0.03868
2016-04-02    0.03540
2016-04-01    0.03380
2016-03-31    0.03192
2016-03-30    0.03362
2016-03-29    0.03418
2016-03-28    0.03484
2016-03-27    0.03104
2016-03-26    0.03248
2016-03-25    0.03174
2016-03-24    0.02910
2016-03-23    0.03238
2016-03-22    0.03294
2016-03-21    0.03752
2016-03-20    0.03782
2016-03-19    0.03490
2016-03-18    0.01306
2016-03-17    0.03152
2016-03-16    0.02950
2016-03-15    0.03398
2016-03-14    0.03662
2016-03-13    0.01556
2016-03-12    0.03678
2016-03-11    0.03248
2016-03-10    0.03212
2016-03-09    0.03322
2016-03-08    0.03330
2016-03-07    0.03596
2016-03-06    0.01394
2016-03-05    0.02538
Name: date_crawled, dtype: float64


2016-04-03    0.03868
2016-03-20    0.03782
2016-03-21    0.03752
Name: date_crawled, dtype: float64


count          50000
unique            34
top       2016-04-03
freq            1934
Name: date_crawled, dtype: o

In [13]:
sort_index_ad_created = autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

top3_ad_created_date = autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_values(ascending=False).head(3)

desc_index_ad_created = autos["ad_created"].str[:10].describe()

print(sort_index_ad_created)
print("\n")
print(top3_ad_created_date)
print("\n")
print(desc_index_ad_created)

2016-04-07    0.00128
2016-04-06    0.00326
2016-04-05    0.01184
2016-04-04    0.03688
2016-04-03    0.03892
               ...   
2015-12-05    0.00002
2015-11-10    0.00002
2015-09-09    0.00002
2015-08-10    0.00002
2015-06-11    0.00002
Name: ad_created, Length: 76, dtype: float64


2016-04-03    0.03892
2016-03-20    0.03786
2016-03-21    0.03772
Name: ad_created, dtype: float64


count          50000
unique            76
top       2016-04-03
freq            1946
Name: ad_created, dtype: object


In [14]:
sort_index_last_seen = autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False)
top3_last_seen = autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_values(ascending=False).head(3)

desc_index_last_seen = autos["last_seen"].str[:10].describe()

print(sort_index_last_seen)
print("\n")
print(top3_last_seen)
print("\n")
print(desc_index_last_seen)

2016-04-07    0.13092
2016-04-06    0.22100
2016-04-05    0.12428
2016-04-04    0.02462
2016-04-03    0.02536
2016-04-02    0.02490
2016-04-01    0.02310
2016-03-31    0.02384
2016-03-30    0.02484
2016-03-29    0.02234
2016-03-28    0.02086
2016-03-27    0.01602
2016-03-26    0.01696
2016-03-25    0.01920
2016-03-24    0.01956
2016-03-23    0.01858
2016-03-22    0.02158
2016-03-21    0.02074
2016-03-20    0.02070
2016-03-19    0.01574
2016-03-18    0.00742
2016-03-17    0.02792
2016-03-16    0.01644
2016-03-15    0.01588
2016-03-14    0.01280
2016-03-13    0.00898
2016-03-12    0.02382
2016-03-11    0.01252
2016-03-10    0.01076
2016-03-09    0.00986
2016-03-08    0.00760
2016-03-07    0.00536
2016-03-06    0.00442
2016-03-05    0.00108
Name: last_seen, dtype: float64


2016-04-06    0.22100
2016-04-07    0.13092
2016-04-05    0.12428
Name: last_seen, dtype: float64


count          50000
unique            34
top       2016-04-06
freq           11050
Name: last_seen, dtype: object


**Observation:**
- The range of the dates in `date_crawled` and `last_seen` columns are all between 5 March. 2016 to 7 April. 2016.
- `ad_created` column contains dates that are earlier (starting from June 2015), but the last day created were also on 07 April. 2016.
- `date_crawled` and `ad_created` columns demonstrate the same dates for top-three frequent date - **3 April 2016, 20 March 2016 and 21 March 2016**. 
- All three columns have the most frequent date on 6 April. 2016.


From this data, it seems that most of the crawlers were on the website between March 2016 to April 2016. This trend can be confirmed in both `data_crawled` and `last_seen` columns.

Also, the same top three frequent dates between `data_crawled` and `ad_created` might imply that the advertisment added has some relationship with when the crawlers crawl the website.



In [15]:
# Have a quick look of the integer columns 
autos[["registration_month", "registration_year"]].head()

registration_month  registration_year
0                   3               2004
1                   6               1997
2                   7               2009
3                   6               2007
4                   7               2003

In [16]:
# Explore the column registration_month
desc_registration_m = autos["registration_month"].describe()
top3_registration_m = autos["registration_month"].value_counts(normalize=True, dropna=False).sort_values(ascending=False).head(3)

print(desc_registration_m)
print("\n")
print(top3_registration_m)

count    50000.000000
mean         5.723360
std          3.711984
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64


0    0.10150
3    0.10142
6    0.08736
Name: registration_month, dtype: float64


In [17]:
zero_registration_month = autos.loc[autos["registration_month"] == 0, "registration_month"]
zero_registration_month.shape

(5075,)

The minimun value in registration_month is 0. But there is no month that is "0". This means that there are missing.
We found that there are 5,075 missing data in `registration_month`.

In [18]:
# Explore the column registration_year
desc_registration_year = autos["registration_year"].describe()
sort_registration_year = autos["registration_year"].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

print(desc_registration_year)
print("\n")
print(sort_registration_year)

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


9999    0.00008
9996    0.00002
9000    0.00004
8888    0.00002
6200    0.00002
         ...   
1800    0.00004
1500    0.00002
1111    0.00002
1001    0.00002
1000    0.00002
Name: registration_year, Length: 97, dtype: float64


**Observations:**
- The max value is 9999 and the min value is 1000 in `registration_year`.
- There are some values that could not be our registration years.

The car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is surely incorrect. At the same time, after some googling, I found that number plates have been in use in Germany since 1906 (check [here](https://www.skoda-storyboard.com/en/models/deciphering-number-plates-germany/)). 

Therefore, we are going to fix the data based on our findings. We will select data only between 1906 and 2016.

In [19]:
registration_y_clean = autos.loc[autos["registration_year"].between(1906, 2016), "registration_year"]

desc_registration_y = registration_y_clean.describe()

sort_registration_y= registration_y_clean.value_counts(normalize=True, dropna=False).sort_values(ascending=False)
top3_registration_y = sort_registration_y.head(3)

print(registration_y_clean)
print("\n")
print(desc_registration_y)
print("\n")
print(top3_registration_y)

 

0        2004
1        1997
2        2009
3        2007
4        2003
         ... 
49995    2011
49996    1996
49997    2014
49998    2013
49999    1996
Name: registration_year, Length: 48028, dtype: int64


count    48028.00000
mean      2002.80351
std          7.31085
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: registration_year, dtype: float64


2000    0.069834
2005    0.062776
1999    0.062464
Name: registration_year, dtype: float64


Though Germany started have the vehicle registration system, almost 75% of the vehicles were registered after 1999 and a large number of car registration were concentrated in the early 2000s.

## Part Five - Explore Price by Brand

In [20]:
brand_desc = autos["brand"].describe()
brand_val_counts = autos["brand"].value_counts(normalize=True)*100

print(brand_desc)
print("\n")
print(brand_val_counts.head(4))


count          50000
unique            40
top       volkswagen
freq           10687
Name: brand, dtype: object


volkswagen       21.374
opel             10.922
bmw              10.858
mercedes_benz     9.468
Name: brand, dtype: float64


**Observations:**
- There are 40 unique vehicle brands in Germany. 
- Four brands, **volkswagen**, **opel**, **bmw** and **mercedes benz** consisted of more than 50% of the vehicle market share. 

Let's check the mean price in each of the four brands.

In [21]:
autos = autos[autos["price"].between(0, 400000)]

brand_price_mean = {}
brand_labels = autos["brand"].value_counts().index

for b in brand_labels:
    mean_price = autos.loc[autos["brand"] == b,"price"].mean().astype(int)
    brand_price_mean[b] = mean_price

vw_price_mean = brand_price_mean["volkswagen"]
opel_price_mean = brand_price_mean["opel"]
bmw_price_mean = brand_price_mean["bmw"]
benz_price_mean = brand_price_mean["mercedes_benz"]

print("The mean price of VOLKSWAGEN is €{:,.2f}".format(vw_price_mean))
print("The mean price of OPEL is €{:,.2f}".format(opel_price_mean))
print("The mean price of BMW is €{:,.2f}".format(bmw_price_mean))
print("The mean price of MERCEDES BENZ is €{:,.2f}".format(benz_price_mean))


The mean price of VOLKSWAGEN is €5,158.00
The mean price of OPEL is €2,845.00
The mean price of BMW is €8,026.00
The mean price of MERCEDES BENZ is €8,389.00


Above analysis shows us that the most expensive price by brand in this four brand is **MERCEDES BENZ**,€8,389, followed by **BMW**, with the mean price of €8,026. The third expensive brand is **VOLKSWAGEN**, €5,158. It is around 1.5 times more expensive than the forth brand **OPEL** with the mean price of €2,845.

In [22]:
# Turn the dictionary to a Series
bmp_series = pd.Series(brand_price_mean).head()
print(bmp_series)

volkswagen       5158
opel             2845
bmw              8026
mercedes_benz    8389
audi             8965
dtype: int64


In [23]:
# Turn the dictionary to a DataFrame
df = pd.DataFrame(bmp_series, columns=["mean_price"]).head()
df

mean_price
volkswagen           5158
opel                 2845
bmw                  8026
mercedes_benz        8389
audi                 8965

## Explore the relationship between odometers and prices

In [50]:
# Check the distribution of the odometers
autos["odometer_km"].value_counts().sort_index(ascending=False)

150000    32416
125000     5169
100000     2168
90000      1757
80000      1436
70000      1230
60000      1164
50000      1025
40000       818
30000       789
20000       784
10000       264
5000        966
Name: odometer_km, dtype: int64

In [54]:
# Find out the mean price in each segment of odometers

odom_m_price = {}
odometer_labels = autos["odometer_km"].value_counts().index

for odometer in odometer_labels:
    mean_price = autos.loc[autos["odometer_km"] == odometer, "price"].mean()
    odom_m_price[odometer] = mean_price

omp_series = pd.Series(odom_m_price).sort_values(ascending=False)
print(omp_series)


10000     19404.571970
20000     17547.154337
30000     16343.263625
40000     15424.191932
50000     13538.203902
60000     12148.767182
70000     10753.006504
80000      9473.653900
90000      8259.381332
100000     7821.486624
5000       7382.675983
125000     6005.635713
150000     3612.810217
dtype: float64


In [57]:
omp_df = pd.DataFrame(omp_series, columns=["Mean Price"])
omp_df

Mean Price
10000   19404.571970
20000   17547.154337
30000   16343.263625
40000   15424.191932
50000   13538.203902
60000   12148.767182
70000   10753.006504
80000    9473.653900
90000    8259.381332
100000   7821.486624
5000     7382.675983
125000   6005.635713
150000   3612.810217

**Observations:**
Roughly the mean price is inversly proportional to the number of mileages. Yet, we found that there is an exceptions. The mean price for the cars with 5,000 mileages are cheaper than those with mileages between 10,000 to 100,000. This finding needs further investigation. Maybe most of the vehicles with 5,000 mileages were cheaper because these vehicles have some problems such as a previous damage.